### PR TITLE
03.Version-cache - framework environment settings

### PR DESCRIPTION
**Pull request by Kalimuthu-Velappan**
It defines and passes the necessary version cache environment variables to the caching framework.
It adds the utils script for shared cache file access.
It also adds the post-cleanup logic for cleaning the unwanted files from the docker/image after the version cache creation.

The following variable controls the version caching feature. By default, it is turned off.
    SONIC_VERSION_CACHE_METHOD = [none/cache]
    SONIC_VERSION_CACHE_SOURCE = /path/to/local/storage

#### Why I did it
It provides a framework for caching build files during the docker build. 

#### How I did it
It uses three staging processes to create a sonic docker image.
   - Stage1:  Creates the sonic docker image
   - Stage2: Copy the intermediate files from the docker image
   - Stage3: Remove the intermediate file from the docker image and creates the final sonic docker image.
   - 
#### How to verify it
Cache files are created under the version cache path after the build.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205

### CODE REVIEW

**Comment by adyeung** [comment]
@xumia @liushilongbuaa submitter has taken the time to split the original code PR to smaller submissions for review, pls help take a look

File: rules/functions [hunk 1]
```diff
@@ -48,6 +48,7 @@ log_green = echo -e "$(GREEN)$(1)$(GRAY)"
 FLUSH_LOG = rm -f $@.log
 
 LOG_SIMPLE = &>> $(PROJECT_ROOT)/$@.log || { [ $$? -eq 0 ] || pushd $(PROJECT_ROOT) > /dev/null ; ./update_screen.sh -e $@ ; popd > /dev/null ; false ; }
+LOG_SAVE = |& $(PROJECT_ROOT)/scripts/process_log.sh $(PROCESS_LOG_OPTION) &>> $(PROJECT_ROOT)/$@.log ; test $${PIPESTATUS[-2]} -eq 0 || { [ $$? -eq 0 ] || pushd $(PROJECT_ROOT) > /dev/null ; ./update_screen.sh -e $@ ; popd > /dev/null ; false ; }
 LOG = < /dev/null |& $(PROJECT_ROOT)/scripts/process_log.sh $(PROCESS_LOG_OPTION) &>> $(PROJECT_ROOT)/$@.log ; test $${PIPESTATUS[-2]} -eq 0 || { [ $$? -eq 0 ] || pushd $(PROJECT_ROOT) > /dev/null ; ./update_screen.sh -e $@ ; popd > /dev/null ; false ; }
```
Comment (xumia) [inline rules/functions:52]: Is it possible to use the same variable form LOG_SAVE and LOG? Looks like the only difference is the "< /dev/null " in the variable LOG.

File: scripts/collect_docker_version_files.sh [hunk 2]
```diff
@@ -18,8 +30,24 @@ export DOCKER_CLI_EXPERIMENTAL=enabled
...
+
+# Save the cache contents from docker build
+LOCAL_CACHE_FILE=target/vcache/${DOCKER_IMAGE_NAME}/cache.tgz
+CACHE_ENCODE_FILE=${DOCKER_PATH}/vcache/cache.base64
+sleep 1; sync ${CACHE_ENCODE_FILE}
+
+# Decode the cache content into gz format
+SRC_VERSION_PATH=files/build/versions
+if [[ -e ${CACHE_ENCODE_FILE} ]]; then
+
```
Comment (xumia) [inline scripts/collect_docker_version_files.sh:47]: removing the empty line?

File: src/sonic-build-hooks/scripts/utils.sh [hunk 3]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+# Lock macro for shared file access
+# Parameters:
+#  $(1) - Lock file name
+#  $(2) - Timeout value
+function FLOCK()
+{
+	local filename=$(dirname $1)
+    local timeout=${2:-360}
```
Comment (xumia) [inline src/sonic-build-hooks/scripts/utils.sh:10]: Could you please fix the tab/space format issue?

File: rules/functions (hunk 1 above)
Comment (Kalimuthu-Velappan) [reply rules/functions]: done

File: scripts/collect_docker_version_files.sh (hunk 2 above)
Comment (Kalimuthu-Velappan) [reply scripts/collect_docker_version_files.sh]: done

File: src/sonic-build-hooks/scripts/utils.sh (hunk 3 above)
Comment (Kalimuthu-Velappan) [reply src/sonic-build-hooks/scripts/utils.sh]: done

**Comment by zhangyanzhao** [comment]
@xumia @liushilongbuaa can you please review the PR? Thanks.

File: slave.mk [hunk 4]
```diff
@@ -558,7 +563,7 @@ $(addprefix $(DEBS_PATH)/, $(SONIC_ONLINE_DEBS)) : $(DEBS_PATH)/% : .platform \
 	if [ -z '$($*_CACHE_LOADED)' ] ; then
 
 		$(foreach deb,$* $($*_DERIVED_DEBS), \
-			{ curl -L -f -o $(DEBS_PATH)/$(deb) $($(deb)_CURL_OPTIONS) $($(deb)_URL) $(LOG) || { exit 1 ; } } ; )
+			{ SKIP_BUILD_HOOK=$($*_SKIP_VERSION) curl -L -f -o $(DEBS_PATH)/$(deb) $($(deb)_CURL_OPTIONS) $($(deb)_URL) $(LOG) || { exit 1 ; } } ; )
```
Comment (xumia) [inline slave.mk:566]: Why we need to skip it? Can we keep the build hook?

File: slave.mk [hunk 5]
```diff
@@ -575,7 +580,7 @@ SONIC_TARGET_LIST += $(addprefix $(DEBS_PATH)/, $(SONIC_ONLINE_DEBS))
 #     SONIC_ONLINE_FILES += $(SOME_NEW_FILE)
 $(addprefix $(FILES_PATH)/, $(SONIC_ONLINE_FILES)) : $(FILES_PATH)/% : .platform
 	$(HEADER)
-	curl -L -f -o $@ $($*_CURL_OPTIONS) $($*_URL) $(LOG)
+	SKIP_BUILD_HOOK=$($*_SKIP_VERSION) curl -L -f -o $@ $($*_CURL_OPTIONS) $($*_URL) $(LOG)
```
Comment (xumia) [inline slave.mk:583]: SAA

File: slave.mk [hunk 6]
```diff
@@ -883,7 +888,7 @@ $(SONIC_INSTALL_WHEELS) : $(PYTHON_WHEELS_PATH)/%-install : .platform $$(addsuff
 	while true; do
 	if mkdir $(PYTHON_WHEELS_PATH)/pip_lock &> /dev/null; then
 ifneq ($(CROSS_BUILD_ENVIRON),y)
-	{ sudo -E pip$($*_PYTHON_VERSION) install $(PYTHON_WHEELS_PATH)/$* $(LOG) && rm -d $(PYTHON_WHEELS_PATH)/pip_lock && break; } || { rm -d $(PYTHON_WHEELS_PATH)/pip_lock && exit 1 ; }
+	{ sudo -E SKIP_BUILD_HOOK=Y pip$($*_PYTHON_VERSION) install $(PYTHON_WHEELS_PATH)/$* $(LOG) && rm -d $(PYTHON_WHEELS_PATH)/pip_lock && break; } || { rm -d $(PYTHON_WHEELS_PATH)/pip_lock && exit 1 ; }
```
Comment (xumia) [inline slave.mk:891]: SAA

File: slave.mk (hunk 5 above)
Comment (Kalimuthu-Velappan) [reply slave.mk:583]: Any particular package that wants to skip versioning selectively can define as below. Eg:
$(BRCM_SAI)_SKIP_VERSION=y

File: slave.mk (hunk 6 above)
Comment (Kalimuthu-Velappan) [reply slave.mk:891]: We don't need to versioning the pip packages that are built as part of the sonic build.

**Review by xumia — APPROVED** [review]

**Comment by xumia** [comment]
@Kalimuthu-Velappan , could you please update the description? Add something in the Why/How/How parts, thanks.

**Comment by Kalimuthu-Velappan** [comment]
It is merged as part of 
https://github.com/sonic-net/sonic-buildimage/pull/12005

**Pull request closed without merging**